scripts/dev/upstream_merge/upstream_merge.sh

```diff
@@ -0,0 +1,206 @@
+      echo ""
+      echo "    Error switching to branch $LOCAL_BASE_BRANCH. Exiting"
+      exit 1
+   fi
+}
+
+add_remote() {
+   UPSTREAM_REPO=$1
+
+   git remote remove $REMOTE_REPO_NAME &> /dev/null || true
+   git remote add $REMOTE_REPO_NAME $UPSTREAM_REPO
+}
+
+sync_remote_branch() {
+   UPSTREAM_BRANCH=$1
+   LOCAL_UPSTREAM_BRANCH=$2
+
+   git branch -D $LOCAL_UPSTREAM_BRANCH &> /dev/null || true
+   if ! `git fetch $REMOTE_REPO_NAME $UPSTREAM_BRANCH:$LOCAL_UPSTREAM_BRANCH &> /dev/null`; then
+      echo ""
+      echo "    Error fetching $REMOTE_REPO_NAME:$UPSTREAM_BRANCH. Exiting"
+      exit 1
+   fi
+}
+
+# Returns 1 if repo should be skipped, 0 if not
+handle_existing_local_branch() {
+   LOCAL_BASE_BRANCH=$1
+   while true; do
+      echo ""
+      echo "    Branch $LOCAL_BRANCH_NAME already exists"
+      read -p "    Delete Branch(d)/Skip Repo(s)/Cancel Merge(c)? " dsc
+      case $dsc in
+         [d]* )
+            if ! `git checkout $LOCAL_BASE_BRANCH &> /dev/null`; then
+               echo ""
+               echo "    Error switching to branch $LOCAL_BASE_BRANCH. Exiting"
+               exit 1
+            fi
+            git branch -D $LOCAL_BRANCH_NAME &> /dev/null;
+            return 0;;
+         [s]* ) return 1;;
+         [c]* ) echo "Exiting"; exit;;
+         * ) echo "    Please answer d/s/c";;
+      esac
+   done < /dev/stdin
+}
+
+# Returns 1 if repo should be skipped, 0 if not
+create_local_branch() {
+   LOCAL_BASE_BRANCH=$1
+   if `git rev-parse --verify $LOCAL_BRANCH_NAME &> /dev/null`; then
+      # Branch already exists
+      if ! handle_existing_local_branch $LOCAL_BASE_BRANCH; then
+         return 1 # Skip repo
+      fi
+   fi
+   if ! `git checkout -b $LOCAL_BRANCH_NAME $LOCAL_BASE_BRANCH &> /dev/null`; then
+      echo ""
+      echo "    Error creating $LOCAL_BRANCH_NAME. Exiting"
+      exit 1
+   fi
+   return 0
+}
+
+# Returns 0 if non empty merge and 1 for empty/no merge
+is_non_empty_merge() {
+   COMMIT_BEFORE_MERGE=$1
+   COMMIT_AFTER_MERGE=`git rev-parse HEAD`
+
+   if [ "$COMMIT_BEFORE_MERGE" == "$COMMIT_AFTER_MERGE" ]; then
+      # No merge commit
+      return 1
+   fi
+
+   if `git diff HEAD~1 HEAD | grep diff &> /dev/null`; then
+      # Non empty merge commit
+      return 0
+   fi
+   # Empty merge commit
+   # This can occur when upstream was previously merged with a squash commit
+   return 1
+}
+
+merge_upstream_branch() {
+   LOCAL_UPSTREAM_BRANCH=$1
+
+   COMMIT_BEFORE_MERGE=`git rev-parse HEAD`
+   if $(git merge $LOCAL_UPSTREAM_BRANCH --signoff -m "Merge latest upstream" &> /dev/null); then
+      if is_non_empty_merge $COMMIT_BEFORE_MERGE; then
+         echo " ... OK"
+      else
+         echo " ... OK (no changes)"
+      fi
+   else
+      echo " ... ERRORS"
+   fi
+}
+
+handle_repo() {
+   LOCAL_REPO=$1
+   UPSTREAM_REPO=$2
+   UPSTREAM_BRANCH=$3
+   LOCAL_BASE_BRANCH=$4
+   LOCAL_UPSTREAM_BRANCH="$LOCAL_UPSTREAM_BRANCH_PREFIX/$UPSTREAM_BRANCH"
```

Comment:
You can use the `local <var_name>=<value>` syntax to create function-scoped variables, so that they don't pollute the global scope.

Reply:
Done